lib/saml2.coffee

```diff
@@ -370,45 +369,78 @@ parse_logout_request = (dom) ->
 
   request
 
+set_option_defaults = (request_options, idp_options, sp_options) ->
+  _.defaults({}, request_options, idp_options, sp_options)
+
 module.exports.ServiceProvider =
   class ServiceProvider
-    constructor: (@issuer, @private_key, @certificate) ->
-    # -- Required
-    # Returns a redirect URL, at which a user can login, and the ID of the request.
-    create_login_url: (identity_provider, assert_endpoint, relay_state..., cb) =>
-      options = relay_state[1]
-      relay_state = relay_state[0]
-
-      { id, xml } = create_authn_request @issuer, assert_endpoint, identity_provider.sso_login_url, options?.force_authn, options?.context, options?.nameid_format
+    # Initializes a service provider given the passed options
+    # @entity_id, @private_key, @certificate, @assert_endpoint can only be set here and are used by exported functions.
+    # Rest of options can be set/overwritten by the identity provider and/or at function call.
+    constructor: (options) ->
+      {@entity_id, @private_key, @certificate, @assert_endpoint} = options
+      @shared_options = _.pick(options, "force_authn", "auth_context", "nameid_format", "sign_get_request", "allow_unencrypted_assertion")
+
+    # Returns:
+    #   Redirect URL at which a user can login
+    #   ID of the request
+    # Params:
+    #   identity_provider
+    #   options
+    #   cb
+    create_login_request_url: (identity_provider, options, cb) ->
+      options = set_option_defaults options, identity_provider.shared_options, @shared_options
+
+      { id, xml } = create_authn_request @entity_id, @assert_endpoint, identity_provider.sso_login_url, options.force_authn, options.auth_context, options.nameid_format
       zlib.deflateRaw xml, (err, deflated) =>
         return cb err if err?
         uri = url.parse identity_provider.sso_login_url
-        if options?.no_signature
+        if options.sign_get_request
           uri.query = SAMLRequest: deflated.toString 'base64'
-          uri.query.RelayState = relay_state if relay_state?
+          uri.query.RelayState = options.relay_state if options.relay_state?
         else
```

Comment:
Just wanted to point one thing out that's probably out of scope for this PR.  This if else loop is a bit confusing given the variable name of `sign_get_request` and function now renamed to `sign_request`.  Irony:
```coffee
if sign_get_request
   //do stuff
else
  sign_request()
```

So we don't call `sign_request` when `sign_get_request` is true, does this seem a bit counterintuitive?